### PR TITLE
Scheduler and window proxy listeners state. Client Rect Poly for IE 11

### PR DESCRIPTION
**Pull request by lynchbomb**
- pinning broccoli-typescript-compiler to `1.0.1`
- https://github.com/linkedin/spaniel/issues/75
- scheduler now handles w.listeners state properly which addresses issues with calling `getBoundingClientRect` on elements not yet within the DOM aka: https://github.com/linkedin/spaniel/issues/76

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -0,0 +1,3744 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (rwjblue) [inline yarn.lock]: Why do we need both a `yarn.lock` and a `package-lock.json`?  Generally speaking you only use one or the other...

Comment (lynchbomb) [reply yarn.lock]: I'll blow this away. I believe its from `yarn link`.

File: src/utils.ts [hunk 2]
```diff
@@ -11,4 +11,16 @@ export function entrySatisfiesRatio(entry: IntersectionObserverEntry, threshold:
   } else {
     return intersectionRatio > threshold || (intersectionRatio === 1 && threshold === 1);
   }
+}
+
+export function getBoundingClientRect(element: Element) {
+  try {
+    return element.getBoundingClientRect();
+  } catch (e) {
+    if (typeof e === 'object' && e !== null) {
```
[the remark this answers is not in the document]
Comment (lynchbomb) [reply src/utils.ts]: Yeah it is odd. <= IE 11 has a botched implementation of `getBoundingClientRect` where if its called on an element *not* yet in the DOM. It throws an unspecified exception.

File: package.json [hunk 3]
```diff
@@ -51,7 +51,7 @@
     "broccoli-merge-trees": "^2.0.0",
     "broccoli-rollup": "^1.3.0",
     "broccoli-string-replace": "^0.1.1",
-    "broccoli-typescript-compiler": "^1.0.1",
+    "broccoli-typescript-compiler": "1.0.1",
```
[the remark this answers is not in the document]
Comment (stefanpenner) [reply package.json]: I think this can safely be unpinned.

File: src/utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (stefanpenner) [reply src/utils.ts]: @lynchbomb we should only catch if it is the exact error IE11 throws here, not all errors.

[the remark this answers is not in the document]
Comment (lynchbomb) [reply src/utils.ts]: Agreed. IE 11 throws the generic "Error" instance in this instance :). Which has a minimal stack trace. What are your thoughts on flagging for the description string "Unspecified" within the error stack?

[the remark this answers is not in the document]
Comment (stefanpenner) [reply src/utils.ts]: sounds good, just prefix that with `typeof e === 'object' && e !== null && /*.. your code above */`.


I did some googling to see if `error.code` is something consistent here, but I was unable to find anything and don't currently have a IE handy

[the remark this answers is not in the document]
Comment (stefanpenner) [reply src/utils.ts]: oops, but `code` i think I meant `number`

[the remark this answers is not in the document]
Comment (stefanpenner) [reply src/utils.ts]: but if it becomes super sketchy to exactly match this error, I can be convinced its not worth it and we can simply `try/catch`...

**Comment by lynchbomb** [comment]
The only other thing I can think of within the `getBoundingClientRect` IE 11 poly. Instead of returning a POJO we should return an instance of `getBoundingClientRect` class. Since it contains many useful prototype methods. However unless we mock the class; it doesn't appear to be a thing we can do in IE 11. So boo.

**Comment by stefanpenner** [comment]
I shouldn't have approved these, there are some issues yet. I should have done more diligence.

**Comment by lynchbomb** [comment]
No prob. I didn’t catch the global scroll event issue either (amongst others). I’ll kick off another PR to mitigate my implementation bork.

**Pull request merged**